docs/proposal/circuit_breaker.md

```diff
@@ -0,0 +1,125 @@
+
+Common scenarios that trigger circuit breakers include:
++ High error rate in the service
++ High latency in the service
++ Exhaustion of service resources
++ Service unavailability
+
+#### Goals
+
++ Support circuit breaker. Kmesh should be able to parse the circuit breaker config from xds stream and enable corresponding circuit breakers.
++ Support circuit breaker. Kmesh should be able to parse the outlier detection config from xds stream and support outlier detection.
++ Add sufficient unit tests to test the correctness of the functions.
+
+### Design Details
+
+#### Implement the state machine of a circuit breaker
+
+<div align="center">
+    <img src="./pics/circuit_breaker_state_machine.png" />
+</div>
+
+We can implement a circuit breaker mechanism to ensure system stability and reliability while minimizing access to faulty services. First, we need to implement a circuit breaker with three states: closed, open, and half-open.
+
+In the closed state, all requests pass through normally and are sent to the target service, even if occasional failures occur, without triggering the circuit breaker. However, when the number of failed requests exceeds a preset threshold, the circuit breaker quickly switches to the open state. In this state, all requests fail immediately, reducing the burden on the system by avoiding sending requests to service that is known to fail.
+
+In the open state, the circuit breaker starts a timeout timer. After the timeout period, the circuit breaker transitions to the half-open state. In the half-open state, the system allows a small number of requests to go through to test whether the target service has recovered. If these requests succeed, the circuit breaker closes again, allowing the service to handle more requests. If the requests fail, the circuit breaker reopens, maintaining protection for the system against the potentially faulty service.
+
+This mechanism effectively monitors the health of the service and dynamically adjusts the state of the circuit breaker based on the actual situation, ensuring the system can quickly respond appropriately to failures. This, in turn, enhances system stability and reliability.
+
+To implement the aforementioned functionality, we can maintain several counters in eBPF to calculate the number of requests reaching each cluster (including the number of successful requests, timed-out requests, failed requests, etc.). Once certain conditions are met, we can trigger the circuit breaker mechanism to reject traffic destined for a particular cluster.
+
+We can use [timers](https://ebpf-docs.dylanreimerink.nl/linux/concepts/timers/) in eBPF to implement the open state of the circuit breaker. When the circuit breaker breaker is open, it will start a timer by the following code:
+
+```c
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 1000);
+    __type(key, int);
```

Comment:
what is the key

Reply:
a circuit is bound to a cluster, maybe we can use the cluster's identifier?